### ISSUE TITLE
Obtuse error after following README when running rocminfo

### ISSUE THREAD

**Issue by aeronth** [nixos-rocm/nixos-rocm#42]
When running `$ nix-shell -p rocminfo --run rocminfo` after cloning the `nixos-rocm` repo and rebuilding according to the README file, I get the following error:

```
error: while evaluating the attribute 'buildInputs' of the derivation 'shell' at 
/nix/store/20ls4hj3bnabbavfxjpzvfqly4wlxs4r-nixpkgs-18.09pre139124.54b90e5f96c/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
The path name '.drv' is invalid: it is illegal to start the name with a period. 
Path names are alphanumeric and can include the symbols +-._?= and 
must not begin with a period. Note: If '.drv' is a source file and you cannot 
rename it on disk, builtins.path { name = ... } can be used to give it an alternative 
name.
```

How can I fix this?

**Comment by acowley** [comment]
The error isn't ringing any bells for me; hopefully someone else will recognize it. Can you show how you installed the overlay? That tends to be a trip hazard for a lot of people.

**Comment by aeronth** [comment]
I installed the overlay by entering the following into `/etc/nixos/configuration.nix`:

```
  # Enable Radeon Open Compute packages for NixOS
  nixpkgs.overlays = [(import /home/aeronth/src/nixos-rocm)];

  boot.kernelPackages = pkgs.linuxPackages_latest;
  hardware.opengl.enable = true;
  hardware.opengl.extraPackages = [ pkgs.rocm-opencl-icd ];
```

`/home/aeronth/src/nixos-rocm` is where the nixos-rocm repo is cloned locally.

Another potentially complicating issue is that I have S9150 cards, which are `gfx701` according to the ROCm page, wasn't sure where that would get added. I added to the `~/.config/nixpkgs/config.nix` file for both `root` and my user account.

I had originally tried with `pkgs.linuxPackages_5_4`, but when it didn't work I zoomed on ahead to the latest kernel to check if it was a kernel issue. Still not working though.

**Comment by acowley** [comment]
Definitely try with a ROCm docker container to see if things work, as I’m not sure about [GPU support](https://github.com/RadeonOpenCompute/ROCm/blob/master/README.md#Hardware-and-Software-Support) for your configuration.

If that doesn’t work, you can potentially get help from AMD, whereas here on NixOS we’re on our own.

I don’t quite follow when you say “I added to the ~/.config/nixpkgs/config.nix file for both root and my user account,” so, just to be sure, beware the [common problem](https://github.com/nixos-rocm/nixos-rocm/issues/40) getting the overlay installed. 

The error message you’re getting is surprising to me, so I’m not sure what’s wrong. I don’t know how a broken `drv` file name is getting involved.

**Comment by aeronth** [comment]
The comment about adding to the `~/.config/nixpkgs/config.nix` file is in reference to this section in the README:

```
To specify GPU compilation targets, your ~/.config/nixpkgs/config.nix 
can include a rocmTargets field that lists GPU targets.
```

I've had this exact configuration working both on Ubuntu and an earlier version of the nixos-ROCm package, so I'm fairly sure the ROCm drivers support the S9150 from AMD. The broken `drv` name file is throwing me off as well, but it feels very `nixos` problem...

**Comment by acowley** [comment]
I agree. Can you inspect the output directory of `nix-build rocminfo`?

**Comment by aeronth** [comment]
Where would I run that? I found the `rocminfo.nix` file in `nixos-rocm/pkgs/development/tool/rocminfo`, but simply typing `nix-build rocminfo` at the command line gives:

```
error: getting status of '/home/aeronth/src/nixos-rocm/rocminfo': No such file or directory
```

Same if I traverse down to the location of the `rocminfo.nix` file.

**Comment by aeronth** [comment]
I found the build outputs in a couple places in the `/nix/store`. I'm trying a `nix-collect-garbage` to see if that helps.

**Comment by acowley** [comment]
On my phone, but I think it’s `nix-build <nixpkgs> -A rocminfo`

**Comment by aeronth** [comment]
OK. The command `nix-build '<nixpkgs>' -A rocminfo` yields an output of:

```
error: The path name '.drv' is invalid: it is illegal to start the name with a period. 
Path names are alphanumeric and can include the symbols +-._?= and must 
not begin with a period. Note: If '.drv' is a source file and you cannot rename it 
on disk, builtins.path { name = ... } can be used to give it an alternative name.
```

**Comment by acowley** [comment]
The `.drv` error still makes me think something is wrong with the overlay installation. You set `rocmTargets` in config.nix; are you also setting your overlays there?

**Comment by aeronth** [comment]
I'm setting the overlay import in `/etc/nixos/configuration.nix`. Do I need to set the `rocmTargets` there as well? Using `config.rocmTargets = [ "gfx701" ];` or something to that effect?

**Comment by aeronth** [comment]
I had a previous overlay for nixos-rocm installed a year or so ago, but removed it. Not sure if that would have left weird artifacts behind...

**Comment by acowley** [comment]
I think that `rocmTargets` needs to be set wherever you are building things. If you do all of your building in `configuration.nix`, i.e., all ROCm packages are in `systemPackages`, then it may be fine to only set that in `configuration.nix`. But if you use `nix-build` and `nix-shell` to kick off builds of ROCm things, then I think you will need that setting in `config.nix`, too.

See the issue I linked above about another user having an issue where they installed the overlay for the system but not their user. It's mentioned in the nixpkgs manual, but it's really subtle and I think more of a pain for regular users like us than NixOS power users who make use of the extra flexibility granted by the system/user distinction.

**Comment by aeronth** [comment]
Got it. That seems quite sensible as an explanation for what might be happening. I will try configuring the settings in both places and check whether that gets everything spinning along properly. Will report back after I have a chance to modify and test.

**Comment by aeronth** [comment]
Thank you for the fast responses!!!

**Comment by aeronth** [comment]
OK. I haven't been able to solve the `nix-shell` issue, could you provide an example of the text for `overlays.nix`? I'm honestly not sure how to write that file, and don't see any ready examples after some extensive Googling.

I did try installing `rocminfo` to my user environment using `nix-env -iA nixos.rocminfo` which works to get the package running, however, I now hit a different issue, when running it the output reads:

```
$ rocminfo
ROCk module is loaded
aeronth is member of video group
hsa api call failure at: /build/source/rocminfo.cc:1102
Call returned HSA_STATUS_ERROR_OUT_OF_RESOURCES: The runtime 
failed to allocate the necessary resources. This error may also occur when the 
core runtime library needs to spawn threads or create internal OS-specific events.
```

Have you seen this error before? Googling shows it as potentially a permissions issue, and it looks like the group for `/dev/kfd` is set to `render` instead of `video`, but adding my user account to both groups doesn't make the error go away.

Thanks!

**Comment by acowley** [comment]
The errors reported by `rocminfo` have been, in my experience, really, extremely unhelpful, so I wouldn't worry about that yet. I think the intent in the program is to naively verify that the system it's running on is setup like a particular developer's machine. What can happen with nix is that we have builders as distinct users that don't have passwords and home directories and such.

Here's an `overlays.nix`. It is a nix expression that evaluates to a list of overlay functions.

```nix
let optOverlay = x: if builtins.pathExists x then [(import x)] else [];
in 
[(import (import ./emacs-overlay/nix/sources.nix).emacs-overlay)
 (import ~/dotfiles/emacs.nix)
 (import ~/dotfiles/nix/mu/overlay.nix)
 (import ~/dotfiles/nix/wordnet.nix)]
++ optOverlay ~/src/nixos-rocm/default.nix
```

Permissions with ROCm can be a problem, but it sounds like you're on top of that.

**Comment by aeronth** [comment]
So if the only overlay I have is `nixos-rocm` would this be something like:

```nix
let optOverlay = x: if builtins.pathExists x then [(import x)] else [];
in
[ (import ~/src/nixos-rocm/default.nix) ]
```

?

I'm trying to figure out why the `nixos-rocm` overlay appears differently in the example above, and this would be the only overlay that I am using.

I get the same error when I use the above best guess for a single overlay syntax:

```
$ nix-shell -p rocminfo --show-trace
error: while evaluating the attribute 'buildInputs' of the derivation 'shell' at /nix/store/20ls4hj3bnabbavfxjpzvfqly4wlxs4r-nixpkgs-18.09pre139124.54b90e5f96c/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:148:11:
The path name '.drv' is invalid: it is illegal to start the name with a period. 
Path names are alphanumeric and can include the symbols +-._?= and 
must not begin with a period. Note: If '.drv' is a source file and you cannot 
rename it on disk, builtins.path { name = ... } can be used to give it an alternative 
name.
```

**Comment by acowley** [comment]
I had just copy/pasted my `overlays.nix`. You don’t need the `optOverlay` definition; it’s a helper that makes it a bit easier for me to use the same nix files on different machines.

We should ask on discourse or IRC to see if the error message means something to others.

**Comment by DieGoldeneEnte** [comment]
Have you solved this already?
The error message `The path name '.drv' is invalid` sounds like you try to build a derivation without a name. It can be created when setting the `name` attribute to `""`. I have no idea how this could happen, since just setting `pname` to `""` is not enough.


This is how I enable the overlay, but your configuration looks fine.
```
let
    nixos-rocm = import ../nixos-rocm;
in
{

  [...]

    nixpkgs = {
        overlays = [ nixos-rocm ];
        config = {
            rocmTargets = [ "gfx906" ];
        };
    };
 }
```
